### PR TITLE
Integrate Detekt plugin

### LINKED ISSUE #1560: [Code health] Integrate kotlin static analysis tool detekt
See https://detekt.dev/ for more details.

### PR DESCRIPTION
**Pull request by shobhitagarwal1612**
Fixes #1560 

There are still a ton of configs that can be enabled. But I'm leaving them in favor of keeping this PR small.

@gino-m PTAL?

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -82,8 +86,21 @@ subprojects {
...
+    source = files("ground/src", "sharedTest/src")
+    allRules = true
+    config = files("${project.rootDir}/config/detekt/detekt.yml")
+    buildUponDefaultConfig = true
+    ignoreFailures = false
+
+    basePath = projectDir
+}
+
+// TODO: Remove java only plugins to improve build speed
```
Comment (gino-m) [inline build.gradle:101]: I think you can remove them straightaway. We shouldn't be writing new Java code, and existing code passes checks.

**Review by gino-m — APPROVED** [review]
Great cleanup, glad to see these checks added! 🧹

File: build.gradle (hunk 1 above)
Comment (shobhitagarwal1612) [reply build.gradle:101]: We still need checkstyle for license header. But rest can be removed

**Pull request merged**